### ISSUE TITLE
False positive MismatchingDocblockReturnType for nullable return statement

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#1436]
Observed: ERROR: MismatchingDocblockReturnType - src/false_mismatch.php:22:8 - Docblock has incorrect return type 'string|null', should be 'string'
Expected: no warning, the return value can be null

```php
<?php

class my_controller {

    const C_NORMAL = 'normal';
    const C_ABNORMAL = 'mobile';

    protected const MAP = [
        'abc' => [
            self::C_NORMAL => 'a',
            self::C_ABNORMAL => 'a',
        ],
        'xyz' => [
            self::C_NORMAL => 'c',
            self::C_ABNORMAL => 'd',
        ],
    ];

    /**
     * @param string $cval
     * @param bool $skipSomething
     * @return string|null
     */
    public static function get_init_snippet(string $cval, bool $skipSomething) : string {
        $variation = $skipSomething ? self::MAP['xyz'][$cval] : null;

        return $variation ? 'something' : null;
    }
}
```

**Comment by muglug** [comment]
check it: `) : string {`

**Comment by muglug** [comment]
Simpler example that makes the error in your code clear: https://psalm.dev/r/d861718856

**Comment by TysonAndre** [comment]
Yeah, just noticed that. Sorry

**Comment by TysonAndre** [comment]
It could also warn on `return $variation ? 'something' : null;` (i.e. discard impossible return types from phpdoc types). but that's a separate request (if MismatchDocblockReturnType is globally suppressed, for example)
